components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaPushSupportFeature.kt

```diff
@@ -0,0 +1,184 @@
+
+    override fun onAuthenticated(account: OAuthAccount, authType: AuthType) {
+        // Register observer when we have an existing account.
+        if (authType != AuthType.Existing) {
+            return
+        }
+
+        accountManager.withConstellation { constellation ->
+            constellation.registerDeviceObserver(observer, lifecycleOwner, autoPause)
+        }
+    }
+
+    override fun onLoggedOut() {
+        // Delete renewal pref.
+        preference(context).edit().remove(LAST_VERIFIED).apply()
+    }
+}
+
+/**
+ * A DeviceConstellation observer to know when we should notify the push feature to begin the registration renewal.
+ */
+internal class ConstellationObserver(
+    context: Context,
+    private val push: PushProcessor,
+    private val verifier: VerificationDelegate = VerificationDelegate(context)
+) : DeviceConstellationObserver {
+
+    private val logger = Logger("ConstellationObserver")
+
+    override fun onDevicesUpdate(constellation: ConstellationState) {
+        val updateSubscription = constellation.currentDevice?.subscriptionExpired ?: false
+
+        // If our subscription has not expired, we do nothing.
+        // If our last check was recent (see: PERIODIC_INTERVAL_MILLISECONDS), we do nothing.
+        if (!updateSubscription || !verifier.shouldVerifyNow()) {
+            return
+        }
+
+        logger.warn("We have been notified that our push subscription has expired; renewing registration.")
+
+        push.renewRegistration()
+
+        verifier.increment()
+    }
+}
+
+/**
+ * A helper that rate limits how often we should notify our servers to renew push registration.
+ *
+ * Implementation notes: This saves the timestamp of our renewal and the number of times we have renewed our
+ * registration within the [PERIODIC_INTERVAL_MILLISECONDS] interval of time.
+ */
+internal class VerificationDelegate(private val context: Context) {
+    @VisibleForTesting
+    internal var innerCount: Int
+    @VisibleForTesting
+    internal var innerTimestamp: Long
+
+    init {
+        val (timestamp, count) = get()
+        innerTimestamp = timestamp
+        innerCount = count
+    }
+
+    /**
+     * Checks whether we're within our rate limiting constraints.
+     */
+    fun shouldVerifyNow(): Boolean {
+        val inInterval = System.currentTimeMillis() - innerTimestamp < PERIODIC_INTERVAL_MILLISECONDS
+        val inRateAcceptance = innerCount <= MAX_REQUEST_IN_INTERVAL
+        val verify = inInterval && inRateAcceptance
+
+        // If it's been PERIODIC_INTERVAL_MILLISECONDS since we last checked, we can reset
+        // out rate limiter and verify now.
+        if (!inInterval) {
+            reset()
+            return true
+        }
+
+        return verify
+    }
+
+    /**
+     * Should be called whenever a successful invocation has taken place and we want to record it.
+     */
+    fun increment() {
+        val count = innerCount + 1
+
+        save("$innerTimestamp;$count")
```

Comment:
I think you want to update `innerTimestamp` to the current time here.

I'm thinking of this as a rolling time window.
```
-----------------------------------------------------------------> time
   |                           |                         |
 last checked                  window                   now
                                 |------------------------|
                                               |
                                        "interval"
```
where `window = currentTime - PERIODIC_INTERVAL_MILLISECONDS`
and `last checked = innerTimestamp`

So when you "increment", you need to bump the "last checked" timestamp. Otherwise, your "inInterval" check may lie to you, since it may think that "last checked" happened longer ago than it actually did.

Reply:
Talked offline: see below issue.